File: packages/sdk/java/packages/android/src/main/java/dev/edgebase/sdk/client/AndroidEdgeBase.java
package dev.edgebase.sdk.client;

import android.app.Activity;

/** Android-only factory that binds a lifecycle/UI context before client use. */
public final class AndroidEdgeBase {
    private AndroidEdgeBase() {
    }

    /**
     * Pass the current Activity when initialization happens after it is resumed
     * so an immediately interactive CAPTCHA has a valid UI host.
     * This overload uses process-memory tokens; anonymous email/phone upgrades
     * fail before network until a DurableTokenStorage is supplied.
     */
    public static ClientEdgeBase client(Activity activity, String url) {
        TurnstileProvider.initialize(activity);
        return new ClientEdgeBase(url);
    }

    /**
     * Create an Android client with app-supplied storage. Anonymous upgrades
     * require the storage to implement {@link DurableTokenStorage}.
     */
    public static ClientEdgeBase client(Activity activity, String url, TokenStorage tokenStorage) {
        TurnstileProvider.initialize(activity);
        return new ClientEdgeBase(url, tokenStorage, null);
    }
}

File: packages/sdk/kotlin/client/src/androidMain/kotlin/dev/edgebase/sdk/client/AndroidEdgeBase.kt
package dev.edgebase.sdk.client

import android.app.Activity

/**
 * Android-specific client factory.
 *
 * Pass the current Activity when creating the SDK from an already-resumed UI.
 * This is the release-safe contract that lets an immediately interactive
 * CAPTCHA attach without relying on hidden Android APIs or a lifecycle replay.
 * The default token storage is process memory. Supply a platform-secure
 * [DurableTokenStorage] before anonymous email/phone account upgrades.
 */
object AndroidEdgeBase {
    /**
     * Create an Android client. [tokenStorage] must implement
     * [DurableTokenStorage] for anonymous replacement-session flows.
     */
    fun client(
        activity: Activity,
        url: String,
        tokenStorage: TokenStorage? = null,
        projectId: String? = null
    ): ClientEdgeBase {
        AndroidActivityTracker.initContext(activity)
        return ClientEdgeBase(url, tokenStorage, projectId)
    }
}

internal actual fun validateClientPlatformInitialization() {
    if (System.getProperty("edgebase.test.allowMissingAndroidActivity") == "true") return
    val activity = AndroidActivityTracker.getCurrentActivity()
    if (activity == null || activity.isFinishing) {
        throw IllegalStateException(
            "EdgeBase Android clients require the current Activity at SDK initialization. " +
                "Use AndroidEdgeBase.client(activity, url, ...) after the Activity is available."
        )
    }
}
